expression/builtin_compare.go

```diff
@@ -383,48 +385,127 @@ func (c *intervalFunctionClass) getFunction(ctx context.Context, args []Expressi
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinIntervalSig{newBaseBuiltinFunc(args, ctx)}
+
+	allInt := true
+	for i := range args {
+		if fieldTp2EvalTp(args[i].GetType()) != tpInt {
+			allInt = false
+		}
+	}
+
+	argTps, argTp := make([]evalTp, 0, len(args)), tpReal
+	if allInt {
+		argTp = tpInt
+	}
+	for range args {
+		argTps = append(argTps, argTp)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpInt, argTps...)
+	var sig builtinFunc
+	if allInt {
+		sig = &builtinIntervalIntSig{baseIntBuiltinFunc{bf}}
+	} else {
+		sig = &builtinIntervalRealSig{baseIntBuiltinFunc{bf}}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinIntervalSig struct {
-	baseBuiltinFunc
+type builtinIntervalIntSig struct {
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinIntervalSig.
+// evalInt evals a builtinIntervalIntSig.
 // See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_interval
-func (b *builtinIntervalSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinIntervalIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	args0, isNull, err := b.args[0].EvalInt(row, sc)
 	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+		return 0, true, errors.Trace(err)
 	}
-	if args[0].IsNull() {
-		d.SetInt64(int64(-1))
-		return
+	if isNull {
+		return -1, false, nil
 	}
-	sc := b.ctx.GetSessionVars().StmtCtx
+	idx, err := b.binSearch(sc, args0, mysql.HasUnsignedFlag(b.args[0].GetType().Flag), b.args[1:], row)
+	return int64(idx), false, errors.Trace(err)
+}
 
-	idx := sort.Search(len(args)-1, func(i int) bool {
-		d1, d2 := args[0], args[i+1]
-		if d1.Kind() == types.KindInt64 && d1.Kind() == d2.Kind() {
-			return d1.GetInt64() < d2.GetInt64()
+// All arguments are treated as integers.
+// It is required that arg[0] < args[1] < args[2] < ... < args[n] for this function to work correctly.
+// This is because a binary search is used (very fast).
+func (b *builtinIntervalIntSig) binSearch(sc *variable.StatementContext, target int64, isUint1 bool, args []Expression, row []types.Datum) (idx int, err error) {
+	i, j, cmp := 0, len(args), false
+	for i < j {
+		mid := i + (j-i)/2
+		v, isNull, err := args[mid].EvalInt(row, sc)
+		if terror.ErrorEqual(err, types.ErrTruncated) {
+			err = sc.HandleTruncate(err)
+		}
+		if err != nil {
+			break
 		}
-		if d1.Kind() == types.KindUint64 && d1.Kind() == d2.Kind() {
-			return d1.GetUint64() < d2.GetUint64()
+		if isNull {
```

Comment:
if `NULL` is treated as `0`, we can not use a binary search like this, since the query can be like `select INTERVAL(100, NULL, NULL, NULL, NULL, NULL, 100);` and mysql will returns `6`

we have to evaluate all the values first, record every non-`NULL` values and its position, then we can do a binary search.